Proposer finalization rewards

nzmdn opened the pull request:
This pull request implements proposer finalization rewards.

The first block of every epoch must contain finalization rewards for the block proposers of the previous epoch. The number of finalization reward outputs equals the epoch length. The scripts to which finalization rewards are sent are taken from the reward (0th) outputs of the previous epoch coinbase transactions. _The reward amount must depend on the percentage of votes included in the previous epoch, but this will be implemented in following PRs_.

See some additional details in #418.

cmihai committed:
Add functional test for finalization rewards

This test checks that the first coinbase in an epoch contains the right
number of reward outputs, that they add up to the right amount, and go
to the right proposer addresses.

Co-authored-by: Azat Nizametdinov <azat@thirdhash.com>
Signed-off-by: Mihai Ciumeica <mihai@thirdhash.com>

cmihai committed:
Fix comment

Signed-off-by: Mihai Ciumeica <mihai@thirdhash.com>

nzmdn committed:
Finalization reward calculation

Co-authored-by: Mihai Ciumeica <mihai@thirdhash.com>
Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

Commit:
Add FinalizationRewardLogic into blockbuilder

Signed-off-by: Matteo Sumberaz <matteo@thirdhash.com>

nzmdn committed:
Test for finalization rewards in the first epoch

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix block reward calculation

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Move RepoMock to test_unite_mocks

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix tests after rebase

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Finalization reward validation WIP

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix functional tests

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix functional tests

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix proposer_balance.py

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Validate non-reward coinbase outputs

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix txvalidationcache_tests

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Remove unused include

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix formatting

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Adjust feature_finalization_rewards

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Check finalization reward maturity when choosing stakeable coins

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Check finalization reward maturity in wallet

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix p2p_unrequested_blocks

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix interface_rest

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix wallet_txn_clone and maturity checks in wallet

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix feature_spend_genesis

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix feature_ltor

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix feature_csv_activation

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix feature_assumevalid

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix p2p_compactblocks

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix feature_versionbits_warning

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Address review comments

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Finalization reward validation after ISD

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Move finalization_reward_logic.cpp from wallet to core

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix number of coinbase outputs check in Consensus::CheckBlockRewards

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Address review comments

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix formatting

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Enable finalization in example_test

Remove unused import in feature_ltor

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Inject finalization::Params to FinalizationRewardLogic

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Remove GetComponent<FinalizationRewardLogic> from WalletExtension

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Address more review comments

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Move reward validation to component

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix example_test.py

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Use CTxOut instead of std::pair<CScript, CAmount>

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Write doc comments

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix functional tests after rebase

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix feature_assumevalid

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix feature_bip68_sequence

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix rpc_propose

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

pep8speaks commented:
Hello @Nizametdinov! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`test/functional/example_test.py`](https://github.com/dtr-org/unit-e/blob/1c08f0460cb41d6d0d5866e2ecdefe68c8f5faad/test/functional/example_test.py):

> [Line 34:1](https://github.com/dtr-org/unit-e/blob/1c08f0460cb41d6d0d5866e2ecdefe68c8f5faad/test/functional/example_test.py#L34): [F401](https://duckduckgo.com/?q=pep8%20F401) 'test_framework.test_framework.DISABLE_FINALIZATION' imported but unused





* In the file [`test/functional/feature_csv_activation.py`](https://github.com/dtr-org/unit-e/blob/1c08f0460cb41d6d0d5866e2ecdefe68c8f5faad/test/functional/feature_csv_activation.py):

> [Line 161:121](https://github.com/dtr-org/unit-e/blob/1c08f0460cb41d6d0d5866e2ecdefe68c8f5faad/test/functional/feature_csv_activation.py#L161): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (151 > 120 characters)

* In the file [`test/functional/feature_finalization_rewards.py`](https://github.com/dtr-org/unit-e/blob/1c08f0460cb41d6d0d5866e2ecdefe68c8f5faad/test/functional/feature_finalization_rewards.py):

> [Line 7:1](https://github.com/dtr-org/unit-e/blob/1c08f0460cb41d6d0d5866e2ecdefe68c8f5faad/test/functional/feature_finalization_rewards.py#L7): [F401](https://duckduckgo.com/?q=pep8%20F401) 'test_framework.test_framework.PROPOSER_REWARD' imported but unused























* In the file [`test/functional/rpc_createmultisig.py`](https://github.com/dtr-org/unit-e/blob/1c08f0460cb41d6d0d5866e2ecdefe68c8f5faad/test/functional/rpc_createmultisig.py):

> [Line 62:121](https://github.com/dtr-org/unit-e/blob/1c08f0460cb41d6d0d5866e2ecdefe68c8f5faad/test/functional/rpc_createmultisig.py#L62): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (129 > 120 characters)







* In the file [`test/functional/test_framework/blocktools.py`](https://github.com/dtr-org/unit-e/blob/1c08f0460cb41d6d0d5866e2ecdefe68c8f5faad/test/functional/test_framework/blocktools.py):

> [Line 87:1](https://github.com/dtr-org/unit-e/blob/1c08f0460cb41d6d0d5866e2ecdefe68c8f5faad/test/functional/test_framework/blocktools.py#L87): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 87:121](https://github.com/dtr-org/unit-e/blob/1c08f0460cb41d6d0d5866e2ecdefe68c8f5faad/test/functional/test_framework/blocktools.py#L87): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (126 > 120 characters)







* In the file [`test/functional/wallet_basic.py`](https://github.com/dtr-org/unit-e/blob/1c08f0460cb41d6d0d5866e2ecdefe68c8f5faad/test/functional/wallet_basic.py):

> [Line 504:121](https://github.com/dtr-org/unit-e/blob/1c08f0460cb41d6d0d5866e2ecdefe68c8f5faad/test/functional/wallet_basic.py#L504): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (134 > 120 characters)

nzmdn committed:
Make changes conform PEP8

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

scravy commented on src/Makefile.am:
```diff
@@ -205,6 +205,7 @@ UNITE_CORE_H = \
   policy/policy.h \
   policy/rbf.h \
   proposer/block_builder.h \
+  proposer/finalization_reward_logic.h \
```
We strive to keep this list alphabetically sorted.

Nizametdinov replied:
Fixed

scravy commented on src/esperanza/walletextension.cpp:
```diff
@@ -13,6 +13,7 @@
 #include <esperanza/finalizationstate.h>
 #include <esperanza/script.h>
 #include <finalization/state_repository.h>
+#include <injector.h>
```
What is this include needed for?

It should not be necessary to include `injector.h` anywhere but in `init.cpp` or in legacy bitcoin code which is not a component. The wallet extension I have to admit is Frankensteins Component, but does have the walletextension_deps to cope with components and dependencies.

Anyways I can not see any particular usage of anything exported by this include.

Nizametdinov replied:
That was a leftover. I began with `GetComponent` and then moved the dependency to `WalletExtensionDeps`.

nzmdn committed:
Fix block generation in feature_bip68_sequence

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Address review comments

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Remove debug log

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

cmihai commented on src/proposer/finalization_reward_logic.h:
```diff
@@ -0,0 +1,71 @@
+  //!
+  //! The first block of every epoch (i.e. previous_block is a checkpoint) must
+  //! contain finalization rewards for the block proposers of the previous
+  //! epoch. The number of finalization reward outputs equals the epoch length.
```
> The number of finalization reward outputs equals the epoch length

As indicated in #418, we should also consider non-optimal finalization, where the dynasty is several epochs long.

Nizametdinov replied:
We decided to pay out rewards after each checkpoint but make rewards mature only after finalization. That means that the number of reward outputs always equals the epoch length.

nzmdn committed:
Remove some leftover code and fix formatting

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

Nizametdinov commented on src/coins.h:
```diff
@@ -64,8 +64,8 @@ class Coin
+    //! \param prevout_index The output index.
     //! \param spend_height The height at which the TxOut is tried to be spent.
-    //! \param prevout_index The output index
     bool IsImmatureCoinBaseReward(const uint32_t prevout_index, const int spend_height) const {
```
#1085

scravy commented:
Needs rebase.

nzmdn committed:
More detailed documentation for finalization_reward_fraction

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

Nizametdinov commented on src/blockchain/blockchain_parameters.h:
```diff
@@ -171,9 +171,12 @@ struct Parameters {
+  //! \brief The fraction of the reward immediately given upon block proposal.
   ufp64::ufp64_t immediate_reward_fraction;
 
+  //! \brief The fraction of the reward given to the proposers after finalization.
```
Done.

Nizametdinov commented on src/proposer/finalization_reward_logic.cpp:
```diff
@@ -0,0 +1,106 @@
+
+    std::vector<CTxOut> result;
+    result.reserve(m_finalization_params->epoch_length);
+    const auto epoch = m_finalization_params->GetEpoch(prev_height);
```
That's right. But the size of the reward will depend on the number of included votes. See "Unit-e Proposers economic model" [here](https://github.com/dtr-org/economics/blob/f75abfc089c91105e192205e4791fccf29558b72/unit_e_economics.ipynb).

castarco replied:
Then, maybe we should change the method name, right?

castarco replied:
> hen, maybe we should change the method name, right?

Ok, I just saw this comment: https://github.com/dtr-org/unit-e/pull/1081/files#r282007048

scravy replied:
Can we resolve this conversation?

Gnappuraz commented on src/staking/block_reward_validator.cpp:
```diff
@@ -15,27 +15,56 @@ namespace staking {
+    if (num_reward_outputs > 1 && !(prev_block.pprev->nStatus & BLOCK_HAVE_DATA)) {
+      // prev_block is a parent block of the snapshot which was used for ISD.
+      // We do not have data for the ancestor blocks of prev_block.
+      // TODO UNIT-E: implement proper validation of finalization rewards after ISD
```
I'd like to have an issue for this. It seems that we cannot check if the rewards are being payed to the correct scripts here, correct?

scravy replied:
Did we create an issue for this?

Gnappuraz commented on test/functional/proposer_balance.py:
```diff
@@ -60,7 +60,9 @@ def test_empty_blocks_balance(self, created_money, nodes):
             coinstake_tx_info = nodes[node_idx].gettransaction(coinstake_tx_id)
 
-            created_money += coinstake_tx_info['details'][0]['amount']
+            # TODO UNIT-E: gettransaction should not return fees for coinbase transactions
```
I would like to have an issue here as well. We don't want to forget this stuff.

scravy replied:
Did we create an issue for this?

Gnappuraz replied:
Not sure, can't find anything.

Gnappuraz reviewed (commented):
Looks good to me, utACK https://github.com/dtr-org/unit-e/pull/1081/commits/5be8a8f18eea23628242724e47c4503acf43377b

nzmdn committed:
Remove unnecessary includes and fix formatting

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix TestChain100Setup after rebase

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Extract maturity logic from wallet.cpp

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Update rpc_getblockstats.json

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

Nizametdinov commented on src/esperanza/walletextension.h:
```diff
@@ -202,6 +202,45 @@ class WalletExtension : public staking::StakingWallet {
+      const CTransaction &tx,
+      const CBlockIndex *block,
+      blockchain::Depth depth,
+      Callable f) const {
```
I used function templates to work with callbacks. But I'm not sure this is the best approach.

scravy replied:
We (and bitcoin) use it in a lot of places, so I for one think this is fine. We have other (like `ForEackStakeableCoin`) functions in this unit that work like this and we have functions like that in `net` and `net_processing`.

nzmdn committed:
Remove duplicate include

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

kostyantyn reviewed (commented):
ConceptACK https://github.com/dtr-org/unit-e/pull/1081/commits/19a3a3b47cc374175fa38da24e2822109aa1d5c0

scravy commented:
`rpc_getblockstats` failed.

I guess you should run it with the `--gen-test-data` argument to update the testdata in `rpc_getblockstats.json` as the block structure changed.

scravy commented on src/esperanza/walletextension.h:
```diff
@@ -202,6 +202,45 @@ class WalletExtension : public staking::StakingWallet {
+      blockchain::Depth depth,
+      Callable f) const {
+    std::size_t reward_offset = 0;
+    if (tx.IsCoinBase() && !block) {
```
maybe wrap both of these `if` clauses in one bigger `if (tx.IsCoinBase())`?

Nizametdinov replied:
Done.

scravy commented on src/esperanza/walletextension.h:
```diff
@@ -202,6 +202,45 @@ class WalletExtension : public staking::StakingWallet {
+  }
+
+  template <typename Callable>
+  void ForEachOutputWithMaturity(
```
I think I would appreciate a comment on both `ForEachMatureOutput` and `ForEachOutputWithMaturity` which especially explains what is the distinction between these two (they sound suspiciously similar)

Nizametdinov replied:
Added comments.

scravy commented on src/test/staking/block_reward_validator_tests.cpp:
```diff
@@ -11,6 +11,33 @@
 
 namespace {
 
+class FinalizationRewardLogicMock : public proposer::FinalizationRewardLogic {
```
Maybe this can be moved to the `unit_e_mocks.h` header file and follow the pattern outlined there?

Nizametdinov replied:
Moved and refactored it.

Gnappuraz commented:
utACK 19a3a3b

castarco commented on src/blockchain/blockchain_parameters.h:
```diff
@@ -171,9 +171,19 @@ struct Parameters {
+  //!
+  //! Proposers receive their reward in two parts. The first part is paid
+  //! immediately in the coinbase transaction of the proposed block. The second
+  //! part is paid in the first block of the following epoch. The amount of the
```
> The second part is paid in the first block of the following epoch

I think this should be rephrased if the second part comes after finalization, as we could enter into a new epoch without having finalization.

Nizametdinov replied:
The second part comes regardless of the finalization. But its size will depend on the number of included votes.

castarco commented on src/esperanza/walletextension.h:
```diff
@@ -202,6 +202,45 @@ class WalletExtension : public staking::StakingWallet {
+    }
+    std::size_t num_immature_rewards = 0;
+    if (tx.IsCoinBase() && depth <= COINBASE_MATURITY) {
+      num_immature_rewards = m_dependencies.GetFinalizationRewardLogic().GetNumberOfRewardOutputs(block->nHeight) + 1;
```
As @scravy commented in the previous function, the two conditions stated in lines `231` & `236` could be wrapped in a shared `if` statement: `if (tx.IsCoinBase()) { ... }`

Nizametdinov replied:
Done.

castarco commented on test/functional/test_framework/blocktools.py:
```diff
@@ -128,6 +133,22 @@ def sign_coinbase(node, coinbase):
+        reward_out = coinbase['vout'][0]
+        reward_script = CScript(hex_str_to_bytes(reward_out['scriptPubKey']['hex']))
+        result.append(CTxOut(FULL_FINALIZATION_REWARD * UNIT, reward_script))
+    return result[::-1]
```
Instead of inverting the result, you could invert the range used for the `for` loop.

Nizametdinov replied:
Replaced with `for i in range(epoch_length)[::-1]`

nzmdn committed:
Update rpc_getblockstats.json

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Address review comments

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Refactor and move FinalizationRewardLogicMock

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Fix formatting

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Add AssertLockHeld to suppress clang error

clang showed the following error:
wallet/wallet.cpp:2577:17: error: calling function 'IsLockedCoin'
  requires holding mutex 'cs_wallet' exclusively
  [-Werror,-Wthread-safety-analysis]
            if (IsLockedCoin(entry.first, i)) {
                ^

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

scravy reviewed (commented):
Where is the `BlockRewardValidator` actually called from, at which step from `CheckBlock`, `ContextualCheckBlock`, to `ConnectBlock` and `CheckStake`?

scravy commented:
The CI build failed due to https://github.com/dtr-org/unit-e/issues/1105

Nizametdinov commented:
> Where is the `BlockRewardValidator` actually called from, at which step from `CheckBlock`, `ContextualCheckBlock`, to `ConnectBlock` and `CheckStake`?

It's called from `ConnectBlock`.

scravy commented on src/proposer/finalization_reward_logic.cpp:
```diff
@@ -0,0 +1,106 @@
+  const Dependency<BlockDB> m_block_db;
+
+  CScript GetRewardScript(const CBlockIndex &index) const {
+    // TODO UNIT-E: implement more efficient reward script retrieval
```
@Nizametdinov Is this comment resolved/acted upon?

Nizametdinov replied:
@scravy I will create an issue when this branch is merged.

scravy commented:
utACK 5320157edbeafb59ef92e7e4939fbed15afe7755

castarco reviewed (commented):
Concept ACK 5320157edbeafb59ef92e7e4939fbed15afe7755

nzmdn committed:
Remove BlockBuilder dependency on FinalizationRewardLogic

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

Nizametdinov commented on src/proposer/block_builder.cpp:
```diff
@@ -94,10 +95,13 @@ class BlockBuilderImpl : public BlockBuilder {
+        m_finalization_reward_logic(finalization_reward_logic) {}
 
   const CTransactionRef BuildCoinbaseTransaction(
+      const CBlockIndex &prev_block,
```
Removed dependency on `FinalizationRewardLogic`. Now the finalization rewards are passed in.

nzmdn committed:
Remove usages of FinalizationState::GetEpochStartHeight

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

Nizametdinov commented on src/wallet/wallet.h:
```diff
@@ -285,7 +285,8 @@ class CMerkleTx
     bool IsInMainChain() const { return GetDepthInMainChain() > 0; }
-    int GetBlocksToRewardMaturity() const;
+    int GetBlocksToRewardMaturity() const { const CBlockIndex *index_ret; return GetBlocksToRewardMaturity(index_ret); }
+    int GetBlocksToRewardMaturity(const CBlockIndex * &index_ret) const;
```
It was used in other places but I refactored those places. Will remove.

nzmdn committed:
Remove unused function

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>

nzmdn committed:
Take into account immediate reward in GetNumberOfRewardOutputs

Signed-off-by: Azat Nizametdinov <azat@thirdhash.com>
